### PR TITLE
Make NSS test-suite and spec compliant

### PR DESCRIPTION
**Pull request by bourgeoa**
With this PR all tests from the test-suite are passing. Considering the following upgrades to be made :
- except the 2 ones relating to PATCH without Write. (see issue https://github.com/solid/specification/issues/220)
- Test suite needs to be upgraded with https://github.com/solid/web-access-control-tests/pull/19 https://github.com/solid/web-access-control-tests/pull/20
- rdflib needs to be upgraded to properly parse JSON-LD https://github.com/linkeddata/rdflib.js/pull/469
- mashlib needs to be upgraded solid/folder-pane https://github.com/solid/folder-pane/pull/17 (in the mean time there is a hack in (ldp.put() to add a content-type)

This is a bit large upgrade. I am sorry for it.

The main changes are : 
- creation of container with PUT (changes in ldp.js)
- resource creation needs acl:accessTo Append and acl:default Write (changes in acl-checker.js, handlers/allow.js, handlers/patch.js)
- delete do not need acl check on folder (changes in ldp-middleware.js)
- content-type is needed (except for create container with PUT. There is no default value.)
- pubsub notification for folder is now send in line with test-suite
 
Some more changes to comply with spec : 
- resource and container cannot only differ with / (changes in handlers/patch.js and ldp.js)

NSS implementation changes : 
- aclCache implement clearing cache with clearAclCache(url) in ldp.delete and ldp.put (changes in ldp.js and acl-checker.js)
- delete of root and root.acl are forbidden
- use the same functions in ldp.js and handlers/patch to create intermediate directories and resource and container doe not use same name differing with /

This PR + NSS 5.6.0 is running on https://solidcommunity.net:8443 test server with mashlib 1.6.5 latest published version working for long chat.

### CODE REVIEW

**Comment by Vinnl** [comment]
It looks like this would also fix #1465, is that correct?

**Comment by bourgeoa** [comment]
@Vinnl yes this solves #1465 however header link is not needed following @csarven comment in https://gitter.im/solid/test-suite
>  btw, that Link header is not necessary for PUT because the effective request URI already has a trailing slash

File: lib/acl-checker.js [hunk 1]
```diff
@@ -40,7 +41,7 @@ class ACLChecker {
 
   // Returns a fulfilled promise when the user can access the resource
   // in the given mode, or rejects with an HTTP error otherwise
```
Comment (TallTed) [inline lib/acl-checker.js:43]:
```suggestion
  // in the given mode; otherwise, rejects with an HTTP error
```

File: lib/acl-checker.js [hunk 2]
```diff
@@ -84,7 +85,19 @@ class ACLChecker {
       // FIXME: https://github.com/solid/acl-check/issues/23
       // console.error(e.message)
     }
-    const accessDenied = aclCheck.accessDenied(acl.graph, resource, directory, aclFile, agent, modes, agentOrigin, trustedOrigins, originTrustedModes)
+    let accessDenied = aclCheck.accessDenied(acl.graph, resource, directory, aclFile, agent, modes, agentOrigin, trustedOrigins, originTrustedModes)
+
+    // For create and update HTTP methods
+    if ((method === 'PUT' || method === 'PATCH' || method === 'COPY') && directory) {
+      // if resource and acl have same parent container,
+      // and resource do not exists then accessTo Append from parent is required
```
Comment (TallTed) [inline lib/acl-checker.js:93]:
```suggestion
      // and resource does not exist, then accessTo Append from parent is required
```

**Review by michielbdejong — APPROVED** [review]
Sounds great!

**Pull request merged**